[OpenMP] Fix race condition in taskgraph with task dependencies

jpinot opened the pull request:
This commit resolves a datarace that could occur when taskgraph with dependencies was enabled alongside task stealing. The problem occur because, during task stealing, the stealing thread could incorrectly add an extra node to the task dependency graph, while the original thread might fail to create the correct node, leading to a corrupted graph and potential execution issues.

**EDIT 07/10/2025**:
The above message is incorrect. The issue comes from the same thread setting a successors node to a node that has already been added, meanwhile the predecessor task have been executed and their dependencies released. Leaving the successor node waiting for the already executed task:


Thread 0: [sets](https://github.com/jpinot/llvm-project/blob/cd9a91992a069416a5a2a82dae66eec7d434c31b/openmp/runtime/src/kmp_taskdeps.cpp#L353) Task1 as successor of Task0
Thread 1: [execute](https://github.com/jpinot/llvm-project/blob/cd9a91992a069416a5a2a82dae66eec7d434c31b/openmp/runtime/src/kmp_tasking.cpp#L1513) Task0
Thread 1: [release](https://github.com/jpinot/llvm-project/blob/cd9a91992a069416a5a2a82dae66eec7d434c31b/openmp/runtime/src/kmp_taskdeps.h#L97) successors of Task0
Thread 0: [re-sets](https://github.com/jpinot/llvm-project/blob/cd9a91992a069416a5a2a82dae66eec7d434c31b/openmp/runtime/src/kmp_taskdeps.cpp#L305) Task1 as successor of Task0
....
// Hangs due to Task1 not being executed because the Task0 dependency was re-set as predecessor after it was executed and their successors released.


Without taskgraph, the nodes are removed once executed; in contrast, when taskgraph is used, the nodes are maintained during and after the execution of the entire region. `kmp_tasking_flags_t::onced` is used to mark a task as executed, but it was incorrectly set. This commits solve the issue [setting](https://github.com/llvm/llvm-project/pull/150880/files#diff-1f0fdb8a559b9c74bb320e9cf3a13de2bf2a15dbaacb1ab9fee92ea1a16216b3R149) `onced` before releasing the dependencies and checks for the same flag when linking(`__kmp_depnode_link_successor`).

jpinot committed:
[OpenMP] Fix datarace with taskgraph dependencies and task stealing

This commit resolves a datarace that could occur when taskgraph with
dependencies was enabled alongside task stealing. The problem occur
because, during task stealing, the stealing thread could incorrectly add
an extra node to the task dependency graph, while the original thread
might fail to create the correct node, leading to a corrupted graph and
potential execution issues.

jprotze commented:
In which case can a different thread insert a node into the dependency graph? The graph should contain only sibling tasks, so how can a different thread create a task that is sibling to tasks in the taskgraph?
Is the problem, that the stealing thread removes nodes from the DAG while the taskgraph is generated?

jpinot committed:
[OpenMP] Fix successor reassignment race in taskgraph

Fix race condition when the same thread re-set a successor node for a
task that had already completed execution and released its dependencies.
This caused the successor to remain waiting indefinitely:

Thread 0 sets Task1 as successor of Task0
Thread 1 executes Task0 and releases successors
Thread 0 re-sets Task1 as successor of Task0
-> Task1 never executes because Task0 was already completed

jpinot committed:
Revert "[OpenMP] Fix datarace with taskgraph dependencies and task stealing"

This reverts commit cd9a91992a069416a5a2a82dae66eec7d434c31b.
